test/development/app-dir/dev-fetch-hmr/dev-fetch-hmr.test.ts

```diff
@@ -0,0 +1,34 @@
+import { nextTestSetup } from 'e2e-utils'
+import { retry } from 'next-test-utils'
+
+import cheerio from 'cheerio'
+
+describe('dev-fetch-hmr', () => {
+  const { next } = nextTestSetup({
+    files: __dirname,
+  })
+
+  it('should retain module level fetch patching', async () => {
+    const html = await next.render('/')
+    expect(html).toContain('monkey patching is fun')
+
+    const magicNumber = cheerio.load(html)('#magic-number').text()
+
+    const html2 = await next.render('/')
+    expect(html2).toContain('monkey patching is fun')
+    const magicNumber2 = cheerio.load(html2)('#magic-number').text()
+    expect(magicNumber).toBe(magicNumber2)
+
+    // trigger HMR
+    await next.patchFile('app/page.tsx', (content) =>
```

Comment:
I understand this test asserts on something the component produces, but that also means that it potentially allows for false positive/negative matches. 

Would you consider modeling a test similarly to how we debugged the issue today? 

1. Introduce a patch by defining a new property on `fetch`.
2. Assert it's there before HMR.
3. Trigger HMR.
4. Assert it's still there.

You can use `Math.random()` for the value of the property, depending on whether the integrity of the value matters (I think it doesn't; as long as Next re-applies the external fetch patch, all is good).

Reply:
This is actually not applicable because we reset the fetch during HMR, and the user land patch is re-applied on top of the original fetch. Instead, I've added a check to ensure that we don't patch over a previous patch.